packages/core/crates/core-strategy/src/promiscuous.rs

```diff
@@ -39,95 +50,159 @@ pub struct PromiscuousStrategy {
-                    to_close.push(address.clone());
+                // Also get channels we have opened with it
+                let channel_with_peer = outgoing_channels
+                    .iter()
+                    .find(|c| c.status == Open && c.destination.eq(&address));
+
+                if let Some(channel) = channel_with_peer {
+                    if quality <= self.config.network_quality_threshold {
+                        // Need to close the channel, because quality has dropped
+                        debug!("new channel closure candidate with {} (quality {})", address, quality);
+                        to_close.push(channel.clone());
+                    } else if channel.balance.lt(&self.config.minimum_channel_balance) {
+                        // Need to re-open channel, because channel stake has dropped
+                        debug!("new channel closure & re-stake candidate with {}", address);
+                        to_close.push(channel.clone());
+                        new_channel_candidates.push((address.clone(), quality));
+                    }
+                } else if quality >= self.config.network_quality_threshold {
+                    // Try to open channel with this peer, because it is high-quality
+                    debug!("new channel opening candidate {} with quality {}", address, quality);
                     new_channel_candidates.push((address.clone(), quality));
                 }
-            } else if quality >= self.network_quality_threshold {
-                // Try to open channel with this peer, because it is high-quality
-                debug!("new channel opening candidate {} with quality {}", address, quality);
-                new_channel_candidates.push((address.clone(), quality));
-            }
 
-            active_addresses.insert(address, quality);
-            network_size += 1;
+                active_addresses.insert(address, quality);
+                network_size += 1;
+            } else {
+                error!("failed to get chain key from a packet key");
+                continue;
+            }
         }
-        self.sma.add_sample(network_size);
+
+        self.sma.write().await.add_sample(network_size);
         info!("evaluated qualities of {} peers seen in the network", network_size);
 
-        if self.sma.get_num_samples() < self.sma.get_sample_window_size() {
+        if self.sma.read().await.get_num_samples() < self.sma.read().await.get_sample_window_size() {
             info!(
                 "not yet enough samples ({} out of {}) of network size to perform a strategy tick, skipping.",
-                self.sma.get_num_samples(),
-                self.sma.get_sample_window_size()
+                self.sma.read().await.get_num_samples(),
+                self.sma.read().await.get_sample_window_size()
             );
-            return StrategyTickResult::new(0, vec![], vec![]);
+            return Ok((to_close, to_open));
```

Comment:
The method should return `Ok(())`

Reply:
Changed back and introduced `collect_tick_decision` method in Promiscuous strategy to get the intermediate result